Add dialog and 2nd order subdivision capability to femmesh2mesh function

goatchurchprime opened the pull request:
This is to address https://github.com/FreeCAD/FreeCAD/issues/15815

I am testing this code against a standard install of FreeCAD without a build environment by checking out the code, changing `class FemWorkbench(Workbench):` to `class GTCHFemWorkbench(Workbench):` and executing:

> freecad -M /home/julian/repositories/FreeCAD/src/Mod/Fem/

The Python code in `_FEMMesh2Mesh(CommandManager).Activated()` now pops up a dialog when it is appropriate (either we have a `Fem::FemResultObject` object or a 2nd order `Fem::FemMeshObject`), and then executes the necessary `FreeCADGui.doCommand`s with the given values.

#### Issue 1
The `FreeCADGui.PySideUic.loadUi()` function requires an absolute path to find the `ui` file.

The common way to do this in the codebase is like: 
> FreeCAD.getHomePath() + "Mod/Fem/Resources/ui/ConstraintCentrif.ui"

But since I am working with a local copy of the workbench code I've had to do it by:
> os.path.join(os.path.split(\_\_file\_\_)[0], "../femmesh/femmesh2mesh.ui")

Also, the build system appears to move the `.ui` files into different directories from where they were created. 


#### Issue 2
My new `femmeshtofacetsSubdiv()` function hasn't been merged into the `femmesh_2_mesh()` and doesn't handle quad meshes.  This will be done later (but before this becomes a formal Pull Request).


#### Issue 3
Should the `Fem::FemResultObject` have an `App::PropertyLink` that points back to its corresponding `Fem::FemMeshObject` so that selecting the FemResultObject is enough to identify the FemMesh that is to be converted?  

Similarly, should the MeshObject be created with a `App::PropertyLink` that points back to its source FemMeshObject?  

These `App::PropertyLink` back-pointers would make more sense than embedding the FemMeshObject as a child of the FemResultObject when you can have multiple results based on the same mesh.

#### Issue 4
The `mesh_2_femmesh()` function called at the end of the operation generates a FemMesh from the mesh triangles for no reason.  This looks like work-in-progress for a new function (converting Meshes in to FemMeshes).  Is this the intention?

goatchurchprime committed:
add in dialog to set the deformation scale and subdivision options in femmesh

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

maxwxyz commented:
@FEA-eng FYI

FEA-eng commented:
Great, this would also fix #11301. I built it from your branch on Ubuntu VM and I get those errors when trying to use the tool for 2nd order (default) Gmsh input mesh:

```
10:42:57  Running the Python command 'FEM_FEMMesh2Mesh' failed:
Traceback (most recent call last):
  File "/home/ubuntu/Documents/freecad_fem_m2m/freecad-source/build/Mod/Fem/femcommands/commands.py", line 661, in Activated
    dialog = FreeCADGui.PySideUic.loadUi(s)
  File "<string>", line 4, in <module>

Cannot open file
```

For CCX_Results_Mesh, it works but in the old way without a pop-up dialog.

Regarding your question about the behavior of the tool in the opposite direction (FEM mesh from regular mesh), it's a result of this PR: https://github.com/FreeCAD/FreeCAD/pull/11492#issuecomment-2245170682 (explained in text document attached there).

Maybe @marioalexis84 could help with your other questions and issues with this PR.

goatchurchprime committed:
moved ui file into corresponding Gui/Resources/ui

goatchurchprime committed:
cmake file to copy the ui file across and commands to look in 2 places for it

goatchurchprime commented:
By carefully watching the build process on old source code (because the current source code doesn't build under nix and I'm not going to be able to fix it till someone who knows what they're doing does it for the next stable release), I was able to see that `src/Mod/Fem/Gui/CMakeLists.txt` contains the instructions for Make to copy the individual ui files from `src/Mod/Fem/Gui/Resources/ui` to `build/Mod/Fem/Resources/ui` (which is a different relative path).  Therefore I have added the following lines into the commands.py file to handle the difference:

```Python
  s = os.path.join(os.path.split(__file__)[0], "../Gui/Resources/ui/femmesh2mesh.ui")
  if not os.path.exists(s):
      print("for FreeCADGui.PySideUic.loadUi, file does not exist: ", s)
      s = os.path.join(os.path.split(__file__)[0], "../Resources/ui/femmesh2mesh.ui")
      print("trying instead file: ", s)
  dialog = FreeCADGui.PySideUic.loadUi(s)
```

The problem is that code development requires a process of rapid editing and rerunning after numerous small changes, and a huge build process that moves those files to other places before they can be used gets in the way.

This should now work for testing purposes.  

For Issue 4, the text document says:

> part 2
> 
> under the Mesh-> "FEM mesh to mesh" option:
> 
> I have also modified "femmesh2mesh.py" to generate a FEM mesh from the surface mesh. The FEM mesh consists of triangular elements. At the end of the "femmesh_2_mesh" function, a call to the "mesh_2_femmesh" function is made and a FEM mesh is generated with the name "Mesh2Fem".
> 

However, this does not explain why anyone would want this to run, since if you just converted a FemMesh to a Mesh, why would you want an extra copy of a FemMesh derived from the new Mesh?

FEA-eng commented:
Ok, now, after building from your branch, I get the dialog for the input mesh (to set the number of subdivisions) but still not the one for the _CCX_Results_Mesh_ (to set the scale factor).

> However, this does not explain why anyone would want this to run, since if you just converted a FemMesh to a Mesh, why would you want an extra copy of a FemMesh derived from the new Mesh?

I think that it's meant to be used in one way only - to quickly generate a FEM mesh from a regular mesh (created in other means than this tool). But let's ask the author of that PR for some information: @mac-the-bike

@chennes Could you approve the workflow so that we see if the tests are working?

goatchurchprime commented:
Are you selecting both the Results and the FemMesh to get that option?

[image: image.png]



On Mon, 12 Aug 2024 at 13:59, FEA-eng ***@***.***> wrote:

> Ok, now, after building from your branch, I get the dialog for the input
> mesh (to set the number of subdivisions) but still not the one for the
> *CCX_Results_Mesh* (to set the scale factor).
>
> However, this does not explain why anyone would want this to run, since if
> you just converted a FemMesh to a Mesh, why would you want an extra copy of
> a FemMesh derived from the new Mesh?
>
> I think that it's meant to be used in one way only - to quickly generate a
> FEM mesh from a regular mesh (created in other means than this tool). But
> let's ask the author of that PR for some information: @mac-the-bike
> <https://github.com/mac-the-bike>
>
> @chennes <https://github.com/chennes> Could you approve the workflow so
> that we see if the tests are working?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/FreeCAD/FreeCAD/pull/15858#issuecomment-2283920200>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAFFLBRQKQRYI4KYEUZ6QW3ZRCWUDAVCNFSM6AAAAABMKWC4ZSVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDEOBTHEZDAMRQGA>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

FEA-eng commented:
> Are you selecting both the Results and the FemMesh to get that option?

Oh, I was selecting only the mesh object, thanks. Now it's working as expected. That's a really great improvement for this tool.

mac-the-bike commented:
This is reply to FEA-eng's and J Todd's comments.

My contribution to femmesh2mesh.py was to add a function to convert the mesh to a femmesh, with the name "Mesh2Fem". This is a mesh of the surface of the structure.
I wrote this function so that I could user my "erase element" stuff to look "inside" a generated mesh.
If you look at: https://forum.freecad.org/viewtopic.php?p=650000

you will see an example, contrived admittedly, of how this feature can be used.

I made the generation of this mesh automatic because I didn't want to get involved with user interfaces.

FEA-eng commented:
@goatchurchprime Is this PR ready? If so, can you undraft it? It won't be merged now as we are still in feature freeze but it could be merged right after the 1.0 release so hopefully really soon.

goatchurchprime commented:
Thanks for the reminder.  Not ready.  Need to get back onto this, and make it work for 4 sided cells.

maxwxyz commented:
@goatchurchprime any news on this PR?

luzpaz commented:
Ping @goatchurchprime

maxwxyz commented:
@marioalexis84 @ickby maybe?
